diffkemp/simpll/passes/SeparateCallsToBitcastPass.cpp

```diff
@@ -0,0 +1,138 @@
+//
+// This file is published under Apache 2.0 license. See LICENSE for details.
+// Author: Petr Silling, psilling@redhat.com
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the definition of the SeparateCallsToBitcastPass pass.
+/// The pass separates bitcasts from calls to bitcast operators by inserting
+/// new bitcast instructions that transform function arguments and the return
+/// value separately.
+///
+//===----------------------------------------------------------------------===//
+
+#include "SeparateCallsToBitcastPass.h"
+
+#include "Utils.h"
+#include <Config.h>
+#include <llvm/IR/Instructions.h>
+#include <llvm/IR/Operator.h>
+
+/// Separate bitcasts from calls to bitcast operators to make the calls
+/// inlinable.
+PreservedAnalyses
+        SeparateCallsToBitcastPass::run(Function &Fun,
+                                        FunctionAnalysisManager &fam) {
+    std::vector<Instruction *> toRemove;
+
+    for (auto &BB : Fun) {
+        for (auto &Inst : BB) {
+            if (auto Call = dyn_cast<CallInst>(&Inst)) {
+                // Find call instructions calling a bitcast
+                // operator that directly corresponds to a function.
+                auto BitCast =
+                        dyn_cast<BitCastOperator>(Call->getCalledValue());
+
+                if (BitCast && isa<Function>(BitCast->getOperand(0))) {
+                    // Get the bitcasted function.
+                    auto srcFun =
+                            dyn_cast<Function>(BitCast->stripPointerCasts());
+
+                    // Ignore the instruction if the number of
+                    // arguments is lower than the number of parameters.
+                    if (Call->getNumArgOperands()
+                        < srcFun->getFunctionType()->getNumParams()) {
+                        continue;
+                    }
+
+                    // Ignore the instruction if the source function returns
+                    // void and the return value is used.
+                    if (srcFun->getReturnType()->isVoidTy()
+                        && Call->getType()->isVoidTy() && !Call->use_empty()) {
+                        continue;
+                    }
+
+                    // Bitcast all arguments to the types of source function
+                    // parameters. If the number of arguments is higher, use the
+                    // remaining arguments without bitcasting as a vargars is
+                    // present in the call instruction.
+                    std::vector<Value *> newArgs;
+                    auto paramType = srcFun->getFunctionType()->param_begin();
+                    auto endParamType = srcFun->getFunctionType()->param_end();
+                    auto arg = Call->arg_begin();
+                    auto endArg = Call->arg_end();
+
+                    while (paramType != endParamType) {
```

Comment:
Why not use a range-based loop?

    for (auto paramType: srcFun->getFunctionType()->params())

Reply:
Agreed (and same for the one below). I was originally using these iterators a bit differently and did not later realise that I can use a range-based loop here, too.